src/js/pages/catalog/PackageDetailTab.js

```diff
@@ -78,15 +84,73 @@ class PackageDetailTab extends mixin(StoreMixin) {
     });
   }
 
-  componentDidMount() {
-    super.componentDidMount(...arguments);
+  isSelectedVersionLoading() {
+    const { version } = this.props.location.query;
+    const cosmosPackage = CosmosPackagesStore.getPackageDetails();
+
+    return cosmosPackage.getVersion() !== version;
+  }
 
+  retrievePackageInfo() {
     const { packageName } = this.props.params;
     const { version } = this.props.location.query;
+    const packageVersions = CosmosPackagesStore.getPackageVersions(packageName);
+
+    // fetch package versions available only if not cached
+    if (packageVersions == null) {
+      CosmosPackagesStore.fetchPackageVersions(packageName);
+    }
+
     // Fetch package description
     CosmosPackagesStore.fetchPackageDescription(packageName, version);
   }
 
+  componentDidMount() {
+    super.componentDidMount(...arguments);
+
+    this.retrievePackageInfo();
+  }
+
+  componentWillReceiveProps(nextProps) {
+    const isLoadingVersions =
+      CosmosPackagesStore.getPackageVersions(nextProps.params.packageName) ==
+      null;
+
+    this.setState({ isLoadingVersions });
+  }
+
+  componentDidUpdate(prevProps) {
```

Comment:
I'm afraid I don't fully understand why we need all these cases. Please try to adjust the implementation to simplify the data flow.

Reply:
This is part of the lifecycle so whenever we have a change (update) we need this otherwise there is no way to know when to change fetch data.

Reply:
I understand then case but would like to reduce the complexity.